Solutions/Marain.ContentManagement.Api.Services/Marain/Cms/Api/Services/Internal/ContentSummariesResponseMapper.cs

```diff
@@ -0,0 +1,75 @@
+﻿// <copyright file="ContentSummariesResponseMapper.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.Cms.Api.Services.Internal
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Menes;
+    using Menes.Hal;
+    using Menes.Links;
+
+    /// <summary>
+    /// Maps <see cref="Content"/> to and from a <see cref="HalDocument"/>.
+    /// </summary>
+    public class ContentSummariesResponseMapper : IHalDocumentMapper<ContentSummaries, ContentSummariesResponseMappingContext>
+    {
+        private readonly IHalDocumentFactory halDocumentFactory;
+        private readonly IOpenApiWebLinkResolver linkResolver;
+        private readonly ContentSummaryResponseMapper contentSummaryMapper;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ContentResponseMapper"/> class.
+        /// </summary>
+        /// <param name="halDocumentFactory">The factory with which to create <see cref="HalDocument"/> instances.</param>
+        /// <param name="linkResolver">The link resolver to build the links collection.</param>
+        /// <param name="contentSummaryMapper">The mapper for individual <see cref="ContentSummary"/> items.</param>
+        public ContentSummariesResponseMapper(
+            IHalDocumentFactory halDocumentFactory,
+            IOpenApiWebLinkResolver linkResolver,
+            ContentSummaryResponseMapper contentSummaryMapper)
+        {
+            this.halDocumentFactory = halDocumentFactory;
+            this.linkResolver = linkResolver;
+            this.contentSummaryMapper = contentSummaryMapper;
+        }
+
+        /// <inheritdoc/>
+        public void ConfigureLinkMap(IOpenApiLinkOperationMap links)
+        {
+            links.MapByContentTypeAndRelationTypeAndOperationId<ContentSummaries>(Constants.LinkRelations.Self, ContentHistoryService.GetContentHistoryOperationId);
+            links.MapByContentTypeAndRelationTypeAndOperationId<ContentSummaries>(Constants.LinkRelations.Next, ContentHistoryService.GetContentHistoryOperationId);
+        }
+
+        /// <inheritdoc/>
+        public HalDocument Map(ContentSummaries resource, ContentSummariesResponseMappingContext context)
+        {
+            IEnumerable<HalDocument> mappedSummaries = resource.Summaries.Select(x => this.contentSummaryMapper.Map(x, context));
+            HalDocument response = this.halDocumentFactory.CreateHalDocumentFrom(new { Summaries = mappedSummaries.ToArray() });
```

Comment:
Lovely use of anonymous types here. You should tell people about it.

Reply:
Note that using `ToArray` almost always causes an extra copy when used against a lazy `IEnumerable<T>`. Lazy enumerables generally don't know how many elements they will return, preventing `ToArray` from being able to pre-allocate an array of the correct size, so LINQ does its thing of allocating an initial size (which I think is actually 4 entries), then doubling it every time it runs out of space, and then when it's finally done, if that doubling caused any spare capacity it makes one more new array of the exact right size, and copies everything into that.

If you're planning to hang onto the array for a long time, this trimming might be worth doing because it means the long-lived object is no bigger than it needs to be. But if you were going to be discarding it fairly soon anyway (which is probably the case here—the next step is likely to be to serialize the thing and then throw it away, right?) then `ToList` is typically a better bet because that doesn't need to perform the trim before it returns, meaning you get one fewer heap allocations. So although an untrimmed `List<T>` sounds like it must necessarily be less efficient than a trimmed `T[]`, in fact the extra allocation required to get an array of exactly the right length causes more work for the GC than just living with the over-sized `List<T>`, and unless something is happening to offset that cost (e.g., you're hanging onto the thing for a long time, so the long term reduction in total memory usage might be worth the initial higher cost involved in getting an array of exactly the right size) then you want `ToList` instead of `ToArray`.